### ISSUE TITLE
[web, desktop]: framework does not move input focus when accessibility focus moves

### ISSUE THREAD

**Issue by mariamhas** [flutter/flutter#83809]
**Issue:** Focus when using JAWs, Narrator, or NVDA only updates by pressing tab key, but doesn't update focus state with Arrow Key or B Key. 

example: pressing down arrow key seems to focus on a different button in the grid (in this case moving from 2 to 10) and screen reader announces button ten, but then when you click tab key the focus actually moves to button3. 

<img width="510" alt="Screen Shot 2021-06-02 at 12 05 40 PM" src="https://user-images.githubusercontent.com/56849473/120538096-e936e180-c39a-11eb-801b-931155449ccc.png">

**Expectation:** should be able to traverse buttons using arrow or B key and hear screen reader announce it and update the focus to the proper element, not just when using tab key. 

Sample code: https://github.com/morrison999/aira-flutter

**Comment by markusaksli-nc** [comment]
Reproducible on the latest master `2.3.0-17.0.pre.180` with Narrator.
<details>
 <summary>flutter doctor -v</summary>

```
[√] Flutter (Channel master, 2.3.0-17.0.pre.180, on Microsoft Windows [Version 10.0.19041.985], locale et-EE)
    • Flutter version 2.3.0-17.0.pre.180 at C:\Development\flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 320f27713b (6 hours ago), 2021-06-03 02:49:03 -0400
    • Engine revision fe6d19b911
    • Dart version 2.14.0 (build 2.14.0-176.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at C:\Users\marku\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\202.7351085\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.7.7)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.7.30621.155
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.2.0)
    • Android Studio at C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\202.7351085
    • Flutter plugin version 57.0.2
    • Dart plugin version 202.8488
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[√] IntelliJ IDEA Ultimate Edition (version 2021.1)
    • IntelliJ at C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\211.7442.40
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19041.985]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 91.0.4472.77
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 91.0.864.37

• No issues found!
```
</details>

**Comment by yjbanov** [comment]
This is still an issue. The problem is that when the engine sends `SemanticsAction.didGainAccessibilityFocus` to the framework most widgets in the framework ignore it. Also, most widgets implement focus features using the `Focus` widget, so an 80% fix might be to fix the `Focus` widget, which should cover most basic widgets. Widgets that implement their own custom focus management (I think `Editable` is one example) would need to implement it explicitly.

**Comment by yjbanov** [comment]
I also checked macOS' native Calculator and Contacts apps and they also move input focus along with accessibility focus, so this is also an issue for Flutter's desktop targets macOS, Windows, and Linux. Added respective labels.

**Comment by gspencergoog** [comment]
I've added a PR to fix this.

**Comment by yjbanov** [comment]
I got some guidance from our accessibility experts. Here's the summary:

> The guidance is to let the screen reader drive the process. The user may configure the screen reader to move keyboard focus along with accessibility focus. Flutter should act in accordance with user preferences. For example, VoiceOver on macOS has the "Synchronize keyboard focus and VoiceOver cursor" setting. If this setting is turned on, Flutter should move the focus, otherwise it should not. Similarly, JAWS and NVDA on Windows implement something called "virtual buffer" which allows the user to disassociate screen reader focus from keyboard focus, and Flutter should respect that. For TalkBack on Android use the user-signaled intent generated by the accessibility service to drive input focus (e.g. using [ACTION_FOCUS](https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.AccessibilityAction#ACTION_FOCUS), or wait until the user activates the edit box using the "double-tap to edit" action).

I have tested manually, and I can definitely see the effect of the "Synchronize keyboard focus and VoiceOver cursor" setting. When it's off, the screen reader becomes detached from keyboard focus. I'm also able to switch modes using Windows screen readers. In one mode I can cause the keyboard focus to follow the screen reader focus, in another, the two are detached.

## Flutter's relationship with a11y APIs

Flutter finds itself in two situations:

1. On iOS, Android, and other native targets Flutter interacts with the screen reader via the native accessibility API. While I have not surveyed every platform, typically it is possible to observe "accessibility focus" signal separately from "keyboard focus". However, not all platforms supply "keyboard focus" in their accessibility API, so native targets subdivide into the following two sub-categories:
  a. Keyboard focus is observable: screen reader may request keyboard focus explicitly. For example, Android accessibility API has [ACTION_FOCUS](https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.AccessibilityAction#ACTION_FOCUS), which is separate from [ACTION_ACCESSIBILITY_FOCUS](https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.AccessibilityAction#ACTION_ACCESSIBILITY_FOCUS).
  b. Keyboard focus is unobservable: the API only supports observing accessibility focus, with keyboard focus behavior supplied using separate means. For example, iOS and macOS use [this approach](https://developer.apple.com/documentation/objectivec/nsobject/uiaccessibilityfocus). While keyboard focus behavior is not communicated via `UIAccessibilityFocus`, global OS settings specify what keyboard focus behavior should be used (TODO: find out the API for getting the settings). The app is expected to use the [accessibilityElementDidBecomeFocused](https://developer.apple.com/documentation/objectivec/nsobject/1615183-accessibilityelementdidbecomefoc) signal, combine it with system VoiceOver options, then decide what to do with the keyboard focus.
2. On the Web Flutter creates HTML elements as proxy for an accessibility tree. In this mode Flutter does not have access to low-level accessibility signals or to screen reader preferences. Instead, the browser interprets screen reader events and settings, and makes the decision whether the input focus should or should not move to the widget (on the web it generates a DOM "focus" event). IOW, on the web we only get the equivalent of Android's `ACTION_FOCUS` event.

## What do we do today?

The status quo in Flutter right now is as follows:

- `ACTION_ACCESSIBILITY_FOCUS` and `accessibilityElementDidBecomeFocused` are implemented as `SemanticsAction.didGainAccessibilityFocus`.
  - What works: this works correctly on mobile, since there does seem to be a way to configure iOS and Android to cause the keyboard focus to be moved in response to these events.
  - What's broken: macOS is broken as Flutter currently does not respect user settings that require keyboard focus to be moved. Worse, this setting is the default, so the majority of users may be relying on it.
- Android `ACTION_FOCUS`: not implemented, likely just broken.
- Web DOM "focus" event: this is currently redirected to `SemanticsAction.didGainAccessibilityFocus`.
  - What works: this partially works. DOM "focus" event actually indicates keyboard focus change (similar to Android's `ACTION_FOCUS`), so accessibility focus is implied. It is therefore not wrong to report it as `SemanticsAction.didGainAccessibilityFocus`.
  - What's broken: the DOM "focus" event is issued by the browser after the browser already interacted with the platform's accessibility API, combined it with user's system settings, and concluded that keyboard focus should be moved. However, since it is reported to the framework via `SemanticsAction.didGainAccessibilityFocus` and the framework does nothing w.r.t. keyboard focus, this is currently broken.

## Options

### 1: Move focus in response to `didGainAccessibilityFocus`

This is what https://github.com/flutter/flutter/pull/142942 does. One small adjustment we could make there is to only move focus if the target platform is desktop or web. It would fully fix the issue on the web, since keyboard and a11y focus are indistinguishable. On native desktop, the situation will improve considerably, as the default expectation is to move keyboard focus. However, if the user configures the screen reader to detach a11y from from keyboard focus, we will not respect that setting.

**This option is the easiest, but not complete.**

### 2: Introduce `SemanticsAction.didGainKeyboardFocus`

Android's `ACTION_FOCUS` and DOM "focus" events will map to this action. macOS will decide between `didGainKeyboardFocus` and `didGainAccessibilityFocus` based on system settings. The framework moves the keyboard focus to the respective widget when receiving this action.

**This option will take longer to implement, but it is complete.**

### 3: Introduce `AccessibilityFeatures.synchronizeKeyboardFocus`

The platform-specific engine will set `synchronizeKeyboardFocus` to true based on the expected platform behavior. One downside of this option is that it's global. It works for web, where we delegate to the browser. It works on macOS where this maps cleanly onto the macOS VoiceOver settings. However, it does not map onto Android cleanly, as `ACTION_ACCESSIBILITY_FOCUS` and `ACTION_FOCUS` can arrive interchangeably, for example when the user switches between touch screen and keyboard (common on Chromebooks).

## Recommendation

My preference is to go for option 2. It is implementable gradually, with each platform engine implementing the new signal independently from others. However, if this bug is severe enough, I would also be in favor of starting with option 1 to make our current behavior less bad, and then take our time to implement option 2.

**Comment by gspencergoog** [comment]
Yes, option 2.1 seems like the best option to me.  Thanks for the awesome write-up and analysis!

**Comment by goderbauer** [comment]
This is a great analysis! Thank you.

I agree that 2 (or maybe 2.1) seem to be the best options. Let's focus on implementing this right instead of doing something hacky (like 1) short term.

Follow-up question about 2.1: To me it looks like Android would support only moving the input focus (without moving the a11y focus) via ACTION_FOCUS. Would we able to implement this with just a bool argument to `didGainAccessibilityFocus`? Sounds like that would give us the option to either only move a11y focus or to move a11y and input focus together, but not restrict it to just input focus. If this is the case, I would prefer option 2 and make input and a11y focus independent of each other.

Small naming nits for 2:

* We should make sure we are consistent with naming. Not sure if this should be called input focus or keyboard focus. I've mostly used the term input focus, but we should check what terminology we use elsewhere (or other platforms use) and stick to that.
* Maybe the action should just be called `inputFocus` without the `didGain` bit as this is more in line with the other `SemanticsAction`s (like tap, scrollUp) etc. to which the framework must react as opposed to `didGainA11yFocus`, which is a purely informative "action" to which the framework (usually) does not react.

For my own understanding:

> this works correctly on mobile, since there does seem to be a way to configure iOS and Android to cause the keyboard focus to be moved in response to these events.

This sentence is missing a "not", right? iOS and Android do *not* have a way to configure this, right?

**Comment by yjbanov** [comment]
> To me it looks like Android would support only moving the input focus (without moving the a11y focus) via ACTION_FOCUS. Would we able to implement this with just a bool argument to didGainAccessibilityFocus?

This is a better question for our Android experts. I searched around Android and Chrome codebases. I wasn't able to dig up the complete `ACTION_FOCUS` story, but here's one example of how Chromium interprets `ACTION_FOCUS`:

https://osscs.corp.google.com/chromium/chromium/src/+/main:content/public/android/java/src/org/chromium/content/browser/accessibility/WebContentsAccessibilityImpl.java;l=1180;drc=02a2b1682256df8e6b7faebe96a9d7198ad006b3

And here's something in Android's rotary service:

https://osscs.corp.google.com/android/platform/superproject/+/master:packages/apps/Car/SystemUI/src/com/android/systemui/car/window/OverlayViewController.java;l=218;drc=199c39aab24f778a3861aef42393a6c301d9732a

> If this is the case, I would prefer option 2 and make input and a11y focus independent of each other.

Actually, there's another reason why 2.1 might be a better option. At least on desktop, both focuses are frequently moved as a result of a single user action. Therefore, processing that one action in a single callback would be less error-prone. The logic in the framework may have to execute conditional logic about what exact state to change depending on the situation. However, if the situation is described via one or two separate events, potentially in an unknown order, the framework-side logic could get complicated. Here's some pseudo-code:

```dart
Semantics(
  onDidGainAccessibilityFocus: (data) {
    if (data.withKeyboard) {
      focusNode.focus();
    } else {
      doSomethingElse();
    }
  },
)
```

If split into two events, we may end up with:

```dart
Semantics(
  onDidGainAccessibilityFocus: () {
    doSomethingElse();
  },
  onDidGainKeyboardFocus: () {
    focusNode.focus();
  },
)
```

This second one is subtly wrong in that it doesn't execute one of the branches but both. It is also too easy to execute them in the opposite order as the order is engine-defined.

> We should make sure we are consistent with naming... I've mostly used the term input focus

Both "input" and "keyboard" SGTM. Input sounds a little more general. It seems input focus is not always controlled by the keyboard. E.g. Android has "rotary" controls (for cars?) that also control focus and input.

> to which the framework must react as opposed to didGainA11yFocus, which is a purely informative "action" to which the framework (usually) does not react.

I'm not sure I'd describe the status quo this way. There's a lot of ad hoc focus management code w.r.t. `did(Gain|Lose)AccessibilityFocus` in the material and cupertino text fields. It seems we have been aware of this issue, fixing it for some widgets on some platforms in a piecemeal fashion.

> This sentence is missing a "not", right?

Oops! Correct, Android/iOS do not have an OS setting for this.

**Comment by goderbauer** [comment]
From the Android API docs it sounds like at least in theory a11y services can send separate and independent requests for ACTION_FOCUS and ACTION_ACCESSIBILITY_FOCUS:

https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.AccessibilityAction#ACTION_FOCUS
https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.AccessibilityAction#ACTION_ACCESSIBILITY_FOCUS

Regarding the complications if those are separate: Do you have a use case in mind where that matters? My hunch is that you should always respond the same to input focus requests no matter how they were received. And optionally on top of that you can do something additional for a11y focus requests. This will mean that we also have to clean up the a11y focus handling in textfields [1]. Instead of having the textfield listen to a11y focus changes and responding to those by moving input focus, the embeders should send a proper input-focus semantics event if that's what the platform requires. The way this is currently handled in textfields is convincing me even more that we should just handle input and a11y focus completely separate in the framework. (Also cc @chunhtai and @justinmc)

[1] https://github.com/flutter/flutter/blob/02d5286e02e4c9f2b80a49a9dee22e8814cf3764/packages/flutter/lib/src/material/text_field.dart#L1411-L1416

**Comment by yjbanov** [comment]
> Regarding the complications if those are separate: Do you have a use case in mind where that matters?

TBH, I'm not sure I understand the use-case for `did(Gain|Lose)AccessibilityFocus` in the framework. Of all usages today, the only thing we're using it for is to _sometimes_ move the _input_ focus. On top of that, on the web, accessibility focus is not observable at all, so if any significant framework functionality relies on it, it will simply not work on the web. AFAICT, the web can only send `did(Gain|Lose)InputFocus`, and it would be a reliable signal for the framework to move the input focus to the respective widget (i.e. we don't need any fancy logic that's currently done in our text fields).

Perhaps we don't need two kinds of focus events, or any flags. Instead, the engine can decide, based on the idioms of the platform, when/if to request an input focus change in the framework, and the framework should just fulfill the request (no platform-specific code needed in the framework).

If so, then what we can do is introduce `did(Gain|Lose)InputFocus` events, then deprecate and remove the `did(Gain|Lose)AccessibilityFocus` ones.

**Comment by goderbauer** [comment]
>  the engine can decide, based on the idioms of the platform, when/if to request an input focus change in the framework, and the framework should just fulfill the request (no platform-specific code needed in the framework).

This is what I am thinking as well.

> deprecate and remove the did(Gain|Lose)AccessibilityFocus ones.

We should still keep did(Gain|Lose)AccessibilityFocus because some platforms (mobile) support it separately from input focus and we added it for a specific a11y use case in a mobile charting library where you want to change chart highlights based on a11y focus _without_ messing with input focus. Though, we should document that moving input focus based on did(Gain|Lose)AccessibilityFocus is likely a smell - if not a bug.

Long story short: I think we should just add a new SemanticsAction that is "please move the input focus to this node", remove the platform-specific a11y/input focus entanglements from textfield, and have all platform embedders send input focus actions as they see fit based on platform conventions.

**Comment by chunhtai** [comment]
I saw one reference where talkback uses action_focus to sync both a11y focus and input focus for controller type input (like tv control)

The voiceaccess also uses it to focus textfield in response of voice command

At least for voiceacess, you will have use case where you only send action_focus but not a11y focus, so I also think 2 would be better.

As for naming, should we just call them onFocus? In Android and flutter, keyboard focus is just called `focus`

**Comment by goderbauer** [comment]
> As for naming, should we just call them onFocus? In Android and flutter, keyboard focus is just called focus. Also, do we need didLoseFoucs for this? I expect not, but I am not sure whether we will need it in the future.

Calling the action `focus` and the callback `onFocus` sounds good to me. We can always add `didLoseFocus` when we need it, I would skip it for now.

**Comment by yjbanov** [comment]
Alight, I'm convinced. Let's introduce `SemanticsAction.focus`.

One potential use-case I see for `SemanticsAction.unfocus` is in multi-view mode, the browser may move the focus from a Flutter widget to a non-Flutter HTML element on the surrounding page. In this case, the engine must communicate to the framework that the widget no longer has focus. However, in that scenario the entire `View` will lose focus, so it may be sufficient just to send the signal about the view, and not the widget? That part should be covered by the work that @tugorez and @gspencergoog are doing.

**Comment by gspencergoog** [comment]
> However, in that scenario the entire View will lose focus, so it may be sufficient just to send the signal about the view, and not the widget? 

Yes, I think that's the case, and I think that the view-level code should do the right thing in that case.

**Comment by yjbanov** [comment]
PR ready for review. PTAL: https://github.com/flutter/engine/pull/53094

**Comment by yjbanov** [comment]
Another PR ready review: https://github.com/flutter/engine/pull/53134 -- switch from .didGain/LoseAccessibilityFocus to .focus

My last PR will be to wire up `SemanticsAction.focus` in the framework. Then we can update @gspencergoog's PR in terms of the new action and fix this issue on the web. Desktop embedders will have to do their fixes separately. Their situation is muddied by the fact that sometimes they already do the focusing via `didGainAccessibilityFocus`. That would need to be undone.

**Comment by yjbanov** [comment]
This PR makes the new `focus` action available through `Semantics` and `CustomPaint`: https://github.com/flutter/flutter/pull/149374

**Comment by yjbanov** [comment]
https://github.com/flutter/engine/pull/53134 should be the final fix for this issue, but it depends on https://github.com/flutter/flutter/pull/142942. I'll wait for @gspencergoog to land that first, then I'll test and land the engine PR, making sure that everything works correct together.

**Comment by goderbauer** [comment]
Thank you @yjbanov and @gspencergoog for driving forward the implementation of this!

After all those PRs land, there is still one clean-up task left: Switch the Desktop embedders to send focus actions and in the Textfield implementation remove the code that ties a11y focus to input focus from the framework. Do we have a tracking issue for that work?

**Comment by yjbanov** [comment]
@goderbauer I'm not aware of one. This issue kind of covers both web and desktop, but I think each desktop platform should be addressed independently. It's probably not practical to fix Windows, Linux, and macOS all in one go. We should file individual issues for those.

**Comment by chunhtai** [comment]
I filed https://github.com/flutter/flutter/issues/149748 for framework
https://github.com/flutter/flutter/issues/149750 for embedding

**Comment by jtmcdole** [comment]
github sync my local fork closed this for _some reason_
